objects/src/accounts/storage/mod.rs

```diff
@@ -121,51 +158,60 @@ impl AccountStorage {
+        layout[usize::from(AccountStorage::SLOT_LAYOUT_COMMITMENT_INDEX)] =
             StorageSlotType::Value { value_arity: 64 };
 
-        // process entries to extract type data
-        let mut entries = items
-            .into_iter()
-            .map(|item| {
-                if item.index == Self::SLOT_LAYOUT_COMMITMENT_INDEX {
-                    return Err(AccountError::StorageSlotIsReserved(item.index));
-                }
-
-                layout[item.index as usize] = item.slot.slot_type;
-                Ok((item.index as u64, item.slot.value))
-            })
-            .collect::<Result<Vec<_>, AccountError>>()?;
+        // The following loop will:
+        //
+        // - Validate the slot and check it doesn't assign a value to a reserved slot.
+        // - Extract the slot value.
+        // - Count the number of maps to validate `maps`.
+        //
+        // It won't detect duplicates, that is later done by the `SimpleSmt` instantiation.
+        //
+        let mut entries = Vec::with_capacity(AccountStorage::NUM_STORAGE_SLOTS);
+        let mut num_maps = 0;
+        for item in items {
+            if item.index == AccountStorage::SLOT_LAYOUT_COMMITMENT_INDEX {
+                return Err(AccountError::StorageSlotIsReserved(item.index));
+            }
+
+            if matches!(item.slot.slot_type, StorageSlotType::Map { .. }) {
+                num_maps += 1;
+            }
+
+            layout[usize::from(item.index)] = item.slot.slot_type;
+            entries.push((item.index.into(), item.slot.value))
+        }
 
         // add layout commitment entry
         entries.push((
-            Self::SLOT_LAYOUT_COMMITMENT_INDEX as u64,
-            *Hasher::hash_elements(&layout.iter().map(Felt::from).collect::<Vec<_>>()),
+            AccountStorage::SLOT_LAYOUT_COMMITMENT_INDEX.into(),
+            *layout_commitment(&layout),
         ));
 
         // construct storage slots smt and populate the types vector.
         let slots = SimpleSmt::<STORAGE_TREE_DEPTH>::with_leaves(entries)
             .map_err(AccountError::DuplicateStorageItems)?;
 
-        // check if the number of provided maps is bigger than the number of slots reserved for maps
-        let count = layout
-            .iter()
-            .filter(|&slot| matches!(slot, StorageSlotType::Map { .. }))
-            .count();
-
-        if maps.len() > count {
-            return Err(AccountError::StorageMapToManyMaps {
-                expected: (count),
-                actual: (maps.len()),
+        if maps.len() > num_maps {
+            return Err(AccountError::StorageMapTooManyMaps {
+                expected: num_maps,
+                actual: maps.len(),
             });
         }
 
+        // validate the map indices
+        for key in maps.keys() {
+            if *key == AccountStorage::SLOT_LAYOUT_COMMITMENT_INDEX {
+                return Err(AccountError::StorageSlotIsReserved(*key));
+            }
+        }
```

Comment:
This check can probably happen on lines 183 - 185 above. Also, I don't think the returned error variant is correct.

Reply:
Should I introduce a new error variant? 

I think `StorageSlotIsReserved` sounds correct. We have the following others

```

pub enum AccountError {
    AccountCodeAssemblerError(AssemblyError),
    AccountCodeNoProcedures,
    AccountCodeTooManyProcedures { max: usize, actual: usize },
    AccountIdInvalidFieldElement(String),
    AccountIdTooFewOnes(u32, u32),
    AssetVaultUpdateError(AssetVaultError),
    DuplicateStorageItems(MerkleError),
    FungibleFaucetIdInvalidFirstBit,
    FungibleFaucetInvalidMetadata(String),
    HexParseError(String),
    InvalidAccountStorageType,
    NonceNotMonotonicallyIncreasing { current: u64, new: u64 },
    SeedDigestTooFewTrailingZeros { expected: u32, actual: u32 },
    StorageSlotInvalidValueArity { slot: u8, expected: u8, actual: u8 },
    StorageSlotIsReserved(u8),
    StorageSlotArrayNotAllowed(u8, StorageSlotType),
    StorageMapNotFound(u8),
    StorageMapTooManyMaps { expected: usize, actual: usize },
    StubDataIncorrectLength(usize, usize),
}
```

Reply:
Actually we already check this in lines 179 for `SlotItems`. I added a check that every map has a corresponding `SlotItem` that stores its root.

Reply:
This should be sufficient as a check

```
            if matches!(item.slot.slot_type, StorageSlotType::Map { .. }) {
                // check that for every map index there is a map in maps
                if !maps.contains_key(&item.index) {
                    return Err(AccountError::StorageMapNotFound(item.index));
                }
                num_maps += 1;
            }
```

- if there is a `item.index` == `255` the function errors in line 179
- if there is a `item` of type `Map` that does not have a `map` in `maps` the functions errors 
- if there are more `map`s in `maps` than there are `items` the functions errors in 205

So there cannot be a `map` with index `255`.

I updated the comments